examples/ctc.dx

```diff
@@ -0,0 +1,134 @@
+'## Connectionist Temporal Classification
+By Alex Graves et alia.
+
+'[Link to paper](https://www.cs.toronto.edu/~graves/icml_2006.pdf).
+
+' This dynamic programming algorithm computes the probability of a particular sequences of labels
+(without pauses included) given another sequence of label probabilities
+(with pauses included), marginalizing over all possible combination of
+pause lengths.
+It's used for training speech-to-text models on unaligned training data.
+
+'Most implementations of CTC compute the marginal logprob
+and its gradients combining a forward and a backward pass.
+However, as mentioned in the original paper, we should only
+need the last step of the forward pass to compute the marginal
+probability, and Dex's autodiff should produce the backward
+pass automatically.  That makes this code much shorter than
+most implementations.
+
+
+def interleave (blank:v) (labels: m=>v) : (m & (Fin 2))=>v =
+  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
+  -- then flattening the pairs back into a single-index table.
+  pairs = for i. [labels.i, blank]
+  for (i, j). pairs.i.j
+
+def prepend (first: v) (seq: m=>v) : (Unit|m)=>v =
+  -- Concatenates a single element to the beginning of a sequence.
+  for idx. case idx
+    Left () -> first
+    Right i -> seq.i
+
+def prepend_and_interleave (blank:v) (seq: m=>v) : 
+  ((Unit | (m & (Fin 2)))=>v) =
+  -- Turns "text" into " t e x t".
+  -- The output of this function has a slightly complicated output type, which
+  -- has size 1 + 2 * (size m).
+  interleaved = interleave blank seq
+  prepend blank interleaved
+
+def clipidx (n:Type) -> (i:Int) : n =
+  -- Returns element at 0 if less than zero.
+  -- Ideally we could have an alternative
+  -- to Fin that just clips the index at its bounds.
```

Comment:
Agreed, although this algorithm also needs to double-decrement, which makes this approach less appealing in this instance.

Reply:
With the above, you can write `i -| 2`. Maybe you mean something else by "double-decrement"?

Reply:
D'oh, I didn't read carefully enough!  Yes, that'd work.

The other thing that'd make this demo nicer is if one could specify a value to return when indexing goes out of bounds.

Reply:
The out-of-bounds case would be really easy to handle if only we had a dual to `inject` (called `project`?):
```hs
project : (n : Type) ?-> (x : Int) : Maybe n = ...
```
Then, you could do "infinite padding" on an array by simply wrapping it in a function:
```hs
pad : (arr : n=>a) (elem : a) : Int->a = \i.
  case project i of
    Just idx -> arr.idx
    Nothing  -> a
```

Reply:
@apaszke  I like that idea.  One thing I want to mention is that I only needed a ```clipdix``` function here because ```select``` doesn't short circuit.  So even though I had switching logic to avoid out-of-bounds indexing, because both sides were evaluated, the unused branch would cause a runtime error unless I wrapped it in ```clipdix```.